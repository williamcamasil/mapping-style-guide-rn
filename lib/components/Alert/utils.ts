import { ViewStyle } from 'react-native';

import { Icons } from '../../assets';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export type AlertVariantType =
  'success' |
  'info' |
  'warning' |
  'error';

export const getVariantColorStyle = (variant?: AlertVariantType): ColorPaletesType => {
  switch (variant) {
    case 'warning':
      return 'feedbackAlert500';
    case 'error':
      return 'feedbackError500';
    case 'info':
      return 'neutralGray600';
    case 'success':
    default:
      return 'feedbackSuccess500';
  }
};

export const getDefaultIcon = (variant?: AlertVariantType) => {
  switch (variant) {
    case 'success':
      return Icons.Default.Success;
    case 'info':
      return Icons.Default.Information;
    case 'warning':
    case 'error':
    default:
      return Icons.Default.Warning;
  }
};

export const getAlertStyle = (
  theme: AppThemeType,
  variant?: AlertVariantType,
): ViewStyle => {

  switch (variant) {
    case 'warning':
      return {
        backgroundColor: theme.colors.feedbackAlert100,
      };
    case 'error':
      return {
        backgroundColor: theme.colors.feedbackError100,
      };
    case 'info':
      return {
        backgroundColor: theme.colors.neutralGray100,
      };
    case 'success':
    default:
      return {
        backgroundColor: theme.colors.feedbackSuccess100,
      };
  }
};
