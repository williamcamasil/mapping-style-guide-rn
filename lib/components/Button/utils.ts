import { ViewStyle } from 'react-native';

import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export type ButtonVariantType =
  'containedPrimary' |
  'containedSecondary' |
  'outlinedPrimary' |
  'outlinedSecondary' |
  'text';

export type ButtonSizeType =
  'large' |
  'medium' |
  'small';

export const getVariantStyle = (
  theme: AppThemeType,
  variant?: ButtonVariantType,
  disabled?: boolean | null,
): ViewStyle => {
  switch (variant) {
    case 'containedSecondary':
      return {
        backgroundColor: disabled ? theme.colors.primary100 : theme.colors.primary200,
        borderWidth: theme.borders.width.none,
      };
    case 'outlinedPrimary':
      return {
        backgroundColor: 'transparent',
        borderWidth: theme.borders.width.thin,
        borderColor: disabled ? theme.colors.primary100 : theme.colors.primary600,
      };
    case 'outlinedSecondary':
      return {
        backgroundColor: 'transparent',
        borderWidth: theme.borders.width.thin,
        borderColor: disabled ? theme.colors.neutralGray100 : theme.colors.neutralGray300,
      };
    case 'text':
      return {
        backgroundColor: disabled ? theme.colors.neutralGray100 : 'transparent',
        borderWidth: theme.borders.width.none,
      };
    case 'containedPrimary':
    default:
      return {
        backgroundColor: disabled ? theme.colors.primary100 : theme.colors.primaryMain,
        borderWidth: theme.borders.width.none,
      };
  }
};

export const getTextColor = (
  variant?: ButtonVariantType,
  disabled?: boolean | null,
): ColorPaletesType => {
  switch (variant) {
    case 'containedSecondary':
      return disabled ? 'primary300' : 'primaryMain';
    case 'outlinedPrimary':
      return disabled ? 'primary300' : 'primaryMain';
    case 'outlinedSecondary':
      return disabled ? 'neutralGray200' : 'neutralGray600';
    case 'text':
      return disabled ? 'neutralGray400' : 'neutralGray500';
    case 'containedPrimary':
    default:
      return disabled ? 'primary300' : 'neutralWhite';
  }
};

export const getSizeStyle = (theme: AppThemeType, size?: ButtonSizeType, hasText?: boolean): ViewStyle => {
  switch (size) {
    case 'large':
      return {
        height: 64,
        minWidth: 64,
        paddingHorizontal: hasText ? theme.spacings.sXL : undefined,
      };
    case 'small':
      return {
        height: 30,
        minWidth: 30,
        paddingHorizontal: hasText ? theme.spacings.sXS : undefined,
      };
    case 'medium':
    default:
      return {
        height: 48,
        minWidth: 48,
        paddingHorizontal: hasText ? theme.spacings.sMedium : undefined,
      };
  }
};

export const getStylesToBlock = (block: boolean | undefined): ViewStyle | null => {
  if (block) return null;

  return {
    flexDirection: 'row',
  };
};
