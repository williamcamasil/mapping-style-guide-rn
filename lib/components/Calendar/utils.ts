import { dateToString } from 'mapping-context-rn';
import type { MarkedDates, Theme } from 'react-native-calendars/src/types';

import { AppThemeType } from '../../theme';

type ExtendedTheme = Theme & {
  'stylesheet.calendar.header': object;
  'stylesheet.day.basic': object;
};

/**
 * Define o tipo de data para o formato yyyy-MM-dd
 */
export type StringDateType = `${number}${number}${number}${number}-${number}${number}-${number}${number}`;

export const DATE_FORMAT = 'yyyy-MM-dd';

export const DAY_SIZE = 42;

export const getDefaultValue = (value?: StringDateType | null) => (
  value ?? dateToString(new Date(), DATE_FORMAT) as StringDateType
);

export const getDisabledDates = (disabledDates?: StringDateType[]) => {
  if (!disabledDates?.length) return {};
  const dates: MarkedDates = {};
  disabledDates.forEach((date: StringDateType) => {
    dates[date] = {
      disabled: true,
      disableTouchEvent: true,
    };
  });
  return dates;
};

export const getCommonTheme = (theme: AppThemeType): ExtendedTheme => ({
  selectedDayBackgroundColor: theme.colors.primaryMain,
  todayTextColor: theme.colors.primaryMain,
  dayTextColor: theme.colors.neutralGray600,
  textDayFontFamily: theme.typography.weights.bold.fontFamily,
  textMonthFontFamily: theme.typography.weights.bold.fontFamily,
  textDayHeaderFontFamily: theme.typography.weights.bold.fontFamily,
  textDayFontWeight: theme.typography.weights.bold.fontWeight,
  textMonthFontWeight: theme.typography.weights.bold.fontWeight,
  textDayHeaderFontWeight: theme.typography.weights.bold.fontWeight,
  textDayFontSize: theme.typography.sizes.small,
  textMonthFontSize: theme.typography.sizes.small,
  textDayHeaderFontSize: theme.typography.sizes.small,
  'stylesheet.calendar.header': {
    dayHeader: {
      marginTop: theme.spacings.sXS,
      marginBottom: theme.spacings.sXS,
      textAlign: 'center',
      fontSize: theme.typography.sizes.small,
      color: theme.colors.primaryMain,
      fontFamily: theme.typography.weights.bold.fontFamily,
      fontWeight: theme.typography.weights.bold.fontWeight,
      flex: 1,
    },
    week: {
      flexDirection: 'row',
      justifyContent: 'space-between',
    },
  },
  'stylesheet.day.basic': {
    base: {
      height: DAY_SIZE,
      width: DAY_SIZE,
      borderBottomLeftRadius: theme.borders.radius.XL,
      borderBottomRightRadius: theme.borders.radius.XL,
      borderTopLeftRadius: theme.borders.radius.XL,
      borderTopRightRadius: theme.borders.radius.XL,
      alignItems: 'center',
      justifyContent: 'center',
    },
  },
});
