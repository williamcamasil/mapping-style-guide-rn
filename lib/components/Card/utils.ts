import { ColorValue, ViewStyle } from 'react-native';

import { AppThemeType } from '../../theme/default';

export type CardSizeType =
  'small' |
  'big';

export type CardVariantType =
  'default' |
  'highlighted50' |
  'highlighted100';

export const getCardSizeStyle = (theme: AppThemeType, size?: CardSizeType): ViewStyle => {
  switch (size) {
    case 'small':
      return {
        width: 135,
        height: 120,
        padding: theme.spacings.sXS,
      };
    case 'big':
    default:
      return {
        minHeight: 128,
        padding: theme.spacings.sSmall,
      };
  }
};

export const getCardShadowStyle = (theme: AppThemeType, showShadow?: boolean): ViewStyle | null => {
  if (showShadow) {
    return theme.shadows.level4;
  }

  return {
    borderWidth: theme.borders.width.thin,
    borderColor: theme.colors.neutralGray200,
  };
};

export const getCardVariantStyle = (theme: AppThemeType, cardType?: CardVariantType): ViewStyle | null => {
  switch (cardType) {
    case 'highlighted50':
      return {
        borderWidth: theme.borders.width.thin,
        borderColor: theme.colors.primary700,
      };
    case 'highlighted100':
      return {
        borderWidth: theme.borders.width.thin,
        borderColor: theme.colors.primary600,
        backgroundColor: theme.colors.primaryMain,
      };
    case 'default':
    default:
      return null;
  }
};

export const getCardCustomBorderStyle = (theme: AppThemeType, borderColor?: ColorValue): ViewStyle | null => {
  if (!borderColor) return null;

  return {
    borderWidth: theme.borders.width.thin,
    borderColor,
  };
};

export const getArrowColor = (theme: AppThemeType, variant?: CardVariantType, arrowColor?: ColorValue): ColorValue => {
  if (arrowColor) return arrowColor;

  if (variant === 'highlighted100') return theme.colors.neutralWhite;

  return theme.colors.neutralGray500;
};
