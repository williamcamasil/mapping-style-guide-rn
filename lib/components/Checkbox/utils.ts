import { CheckBoxImgPropsType } from '../../assets/images/CheckboxImg';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export type CheckboxVariantType = 'default' | 'secondary';

export const getTextColor = (error?: boolean, variant?: CheckboxVariantType, disabled?: boolean): ColorPaletesType => {
  if (disabled) {
    if (variant === 'secondary') return 'secondary500';
    return 'neutralGray300';
  }

  if (error) {
    if (variant === 'secondary') return 'feedbackError300';
    return 'feedbackError400';
  }

  if (variant === 'secondary') return 'secondary200';

  return 'neutralGray700';
};

const getBackgroundColorStyle = (theme: AppThemeType, variant?: CheckboxVariantType) => {
  if (variant === 'secondary') return theme.colors.secondary700;

  return theme.colors.neutralGray100;
};

const getCheckedBorderColorStyle = (theme: AppThemeType, variant?: CheckboxVariantType) => {
  if (variant === 'secondary') return theme.colors.secondary600;

  return theme.colors.neutralGray200;
};

const getColorStyle = (theme: AppThemeType, error?: boolean, variant?: CheckboxVariantType) => {
  if (error) {
    if (variant === 'secondary') return theme.colors.feedbackError300;
    return theme.colors.feedbackError400;
  }

  return theme.colors.primaryMain;
};

const getDisabledBorderColorStyle = (theme: AppThemeType, variant?: CheckboxVariantType) => {
  if (variant === 'secondary') return theme.colors.secondary500;

  return theme.colors.neutralGray200;
};

export const getBoxStyle = (
  theme: AppThemeType,
  error?: boolean,
  disabled?: boolean,
  checked?: boolean,
  variant?: CheckboxVariantType,
): CheckBoxImgPropsType => {
  if (disabled) {
    return (
      {
        backgroundColor: 'transparent',
        borderColor: getDisabledBorderColorStyle(theme, variant),
        borderWidth: theme.borders.width.thin,
        checkColor: theme.colors.neutralGray200,
        checked,
      }
    );
  }

  if (checked) {
    return (
      {
        backgroundColor: getColorStyle(theme, error, variant),
        borderColor: getCheckedBorderColorStyle(theme, variant),
        borderWidth: theme.borders.width.thick,
        checkColor: theme.colors.neutralWhite,
        checked,
      }
    );
  }

  return (
    {
      backgroundColor: getBackgroundColorStyle(theme, variant),
      borderColor: getColorStyle(theme, error, variant),
      borderWidth: theme.borders.width.thin,
      checkColor: theme.colors.neutralGray100,
      checked,
    }
  );
};
