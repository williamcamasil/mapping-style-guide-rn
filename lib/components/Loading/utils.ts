import { ColorValue } from 'react-native';

import Color from 'color';

import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export function getCircleBackgroundColor(
  theme: AppThemeType,
  color: ColorPaletesType,
): ColorValue | undefined {
  return theme.colors[color];
}

export function getCircleBorderColor(
  theme: AppThemeType,
  color: ColorPaletesType,
): ColorValue | undefined {
  const colorValue = theme.colors[color];

  return Color(colorValue).alpha(0.6).hsl().toString();
}
