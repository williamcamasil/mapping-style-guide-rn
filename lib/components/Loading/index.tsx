import React, { useEffect, useRef } from 'react';
import { StyleSheet, View, Animated } from 'react-native';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';
import { getCircleBorderColor, getCircleBackgroundColor } from './utils';

export type LoadingPropsType = {
  theme: AppThemeType;
  circleSize?: number;
  color: ColorPaletesType;
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  outCircle: {
    borderStyle: 'solid',
    overflow: 'hidden',
    marginHorizontal: 4,
  },
  circle: {
    width: '100%',
    height: '100%',
  },
});

const Loading: React.FC<LoadingPropsType> = ({ circleSize, color, theme }) => {
  const circleAnimation = useRef(new Animated.Value(0));

  const styleOutCircle = useViewStyles(
    () => [
      styles.outCircle,
      {
        borderColor: getCircleBorderColor(theme, color),
        width: circleSize,
        height: circleSize,
        borderRadius: circleSize && circleSize / 2,
        borderWidth: theme.borders.width.hairlene,
      },
    ],
    [circleSize, color, theme],
  );

  const styleCircle = useViewStyles(
    () => [
      styles.circle,
      {
        backgroundColor: getCircleBackgroundColor(theme, color),
        borderRadius: circleSize && circleSize / 2,
      },
    ],
    [theme, color, circleSize],
  );

  useEffect(() => {
    const animation = Animated.loop(
      Animated.timing(circleAnimation.current, {
        toValue: 50,
        duration: 1000,
        useNativeDriver: true,
      }),
    );
    animation.start();
    return () => {
      animation.stop();
    };
  }, []);

  const circleStyle1 = useViewStyles(() => [
    styleCircle,
    {
      opacity: circleAnimation.current.interpolate({
        inputRange: [0, 25, 50],
        outputRange: [1, 0.5, 0],
      }),
    }], [styleCircle]);

  const circleStyle2 = useViewStyles(() => [
    styleCircle,
    {
      opacity: circleAnimation.current.interpolate({
        inputRange: [0, 25, 50],
        outputRange: [0, 1, 0.5],
      }),
    }], [styleCircle]);

  const circleStyle3 = useViewStyles(() => [
    styleCircle,
    {
      opacity: circleAnimation.current.interpolate({
        inputRange: [0, 25, 50],
        outputRange: [0.5, 0, 1],
      }),
    }], [styleCircle]);

  return (
    <View style={styles.container}>
      <View style={styleOutCircle}>
        <Animated.View style={circleStyle1} />
      </View>
      <View style={styleOutCircle}>
        <Animated.View style={circleStyle2} />
      </View>
      <View style={styleOutCircle}>
        <Animated.View style={circleStyle3} />
      </View>
    </View>
  );
};

Loading.defaultProps = {
  circleSize: 8,
};

export default withTheme(Loading);
