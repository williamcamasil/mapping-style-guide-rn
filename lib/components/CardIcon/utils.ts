import { ColorValue } from 'react-native';

import { AppThemeType } from '../../theme/default';
import { CardVariantType } from '../Card/utils';

export const getIconColor = (
  theme: AppThemeType,
  customColor?: ColorValue,
  variant?: CardVariantType,
): ColorValue | string => {
  if (customColor) return customColor;

  if (variant === 'highlighted100') {
    return theme.colors.neutralWhite;
  }

  return theme.colors.primaryMain;
};
