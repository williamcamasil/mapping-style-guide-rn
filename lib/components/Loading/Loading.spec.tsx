import React from 'react';
import { StyleSheet, View as RNView, Animated as RNAnimated } from 'react-native';

import { render } from '@testing-library/react-native';

import Loading from '.';
import { DefaultTheme } from '../../theme';
import { getCircleBorderColor, getCircleBackgroundColor } from './utils';

describe('Loading', () => {

  it('renders correctly', () => {
    render(<Loading circleSize={8} color="neutralWhite" />);
  });

  it('Have a property size in a circle', () => {
    const testRenderer = render(<Loading circleSize={8} color="neutralWhite" />);

    const nativeInstance = testRenderer.UNSAFE_getAllByType(RNView);
    const nativeStyle = StyleSheet.flatten(nativeInstance[1].props.style);

    expect(nativeStyle).toHaveProperty('width', 8);
    expect(nativeStyle).toHaveProperty('height', 8);
  });

  it('Have a property color', () => {
    const testRenderer = render(<Loading circleSize={8} color="neutralWhite" />);

    const nativeInstance = testRenderer.UNSAFE_getAllByType(RNAnimated.View);
    const nativeStyle = StyleSheet.flatten(nativeInstance[0].props.style);

    expect(nativeStyle).toHaveProperty('backgroundColor', '#FFFFFF');
  });

  it('Unmount component', () => {
    const { unmount } = render(<Loading circleSize={8} color="neutralWhite" />);
    unmount();
  });
});

describe('Utils', () => {
  it('getCircleBackgroundColor', () => {
    const colorString = getCircleBackgroundColor(DefaultTheme, 'primaryMain');
    expect(colorString).toEqual(DefaultTheme.colors.primaryMain);
  });

  it('getCircleBorderColor', () => {
    const colorString = getCircleBorderColor(DefaultTheme, 'primaryMain');
    expect(colorString).toEqual('hsla(19.19999999999999, 84.2%, 49.6%, 0.6)');
  });
});

describe('Loading snapshot', () => {
  it('default', () => {
    const tree = render(<Loading circleSize={8} color="neutralWhite" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
