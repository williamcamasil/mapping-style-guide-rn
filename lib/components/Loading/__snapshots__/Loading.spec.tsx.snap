// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading snapshot default 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "borderColor": "hsla(0, 0%, 100%, 0.6)",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "height": 8,
        "marginHorizontal": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 4,
          "height": "100%",
          "opacity": 1,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      {
        "borderColor": "hsla(0, 0%, 100%, 0.6)",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "height": 8,
        "marginHorizontal": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 4,
          "height": "100%",
          "opacity": 0,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      {
        "borderColor": "hsla(0, 0%, 100%, 0.6)",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "height": 8,
        "marginHorizontal": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 4,
          "height": "100%",
          "opacity": 0.5,
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;
